### PR TITLE
Consolidate duplicates from common sources

### LINKED ISSUE #163: Consolidate users's avatar duplicates
Following-up on #141, as described in https://github.com/openzim/sotoki/pull/162#issuecomment-660426100 a lot of users customized their avatar on stack-exchange by using third party providers.

It means that we have profile image url for them but a good amount of them are still duplicates.

@satyamtg identified some of those third party but fixing it will require custom logic for each.

My suggestion would be to wait for our stack-overflow run to complete and decide strategy then with
* number of actual duplicates
* repartition of those across domains/services
* weight of those duplicates.

### PR DESCRIPTION
**Pull request by satyamtg**
This reduces the number of duplicate images by managing some identified types of duplicate images from several URLs mentioned in https://github.com/openzim/sotoki/pull/162#issuecomment-660426100 . Apart from this, it also removes the identicons that come from gravatar with similar patterns as what we generate on the JS side taking reference from the following - https://en.gravatar.com/site/implement/images/

The following new changes are made -
- Bump `zimscraperlib` and `lxml` versions
- `output` is renamed to `output_dir` and made global for better access
- `redirect_file` is made global to avoid passing it everywhere
- redirects are now written using a context manager
- `get_response_headers()` is now an independent function as it is used by a new function
- function `get_redirection_to_duplicate_or_skip()` is added - It checks the URL for various types of duplicates and return a tuple containing name of the file to which image can be redirected and whether we can skip the download
- `process_potential_duplicate()` is added and uses the aforementioned function's result to either write a redirect to the image or to skip the download. It returns True in case it processes the image as a duplicate or False in case the image is not one of known duplicate types. It's return value is used by `download_image()` to continue downloading the image or return without downloading
- All types of duplicated images are saved in common_images directory in the build folder with unique naming scheme

In short, the following is done for each image URL type -
- gravatar.com - do not download for identicon type (as we can generate), download the blank and the other two types of default images only once and write redirection
- googleusercontent - download potentially duplicates only once and write redirection
- blend-exchange - download potentially duplicates only once and write redirection

For stats, I tried on `3dprinting.meta.stackexchange.com` with both this fix and without this fix, and the amount of identicons reduced by a fair amount. Without this, it was around 700 identicons whereas after this, it a little less than 500 identicons, which is roughly 28% decrease in number of downloaded identicons.

### CODE REVIEW

File: requirements.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 Jinja2==2.11.1
-lxml==4.5.0
+lxml==4.5.2
```
Comment (rgaudin) [inline requirements.txt:2]: We should not rely on fixed version unless necessary ; particularly for this which is shared with scraperlib.

File: sotoki/sotoki.py [hunk 2]
```diff
@@ -890,13 +882,153 @@ def get_response_headers(url):
...
+            )
+
+        # update query with size = 128 (its the size in which we download identicons)
+        query.update({"s": "128"})
+        url_parts[4] = urllib.parse.urlencode(query)
+        url_with_size = urllib.parse.urlunparse(url_parts)
+
+        # check if potentially a duplicate default image at size = 128
+        headers = get_response_headers(url_with_size)
+        if headers.get("content-length") == "4268":
```
Comment (rgaudin) [inline sotoki/sotoki.py:950]: Where's this number coming from? How do we keep it updated overtime?
We should add a comment here

File: sotoki/sotoki.py [hunk 3]
```diff
@@ -890,13 +882,153 @@ def get_response_headers(url):
...
+            return (
+                download_and_return_name(
+                    url, "gravatar_default1", convert_png, resize, ext
+                ),
+                False,
+            )
+        elif headers.get("content-length") == "3505":
```
Comment (rgaudin) [inline sotoki/sotoki.py:957]: Same here

File: sotoki/sotoki.py [hunk 4]
```diff
@@ -890,13 +882,153 @@ def get_response_headers(url):
...
+                ),
+                False,
+            )
+
+    return None, None
+
+
+def process_potential_duplicate(url, fullpath, convert_png, resize):
+    """ write a redirection for duplicate files or skip download if we can generate similar identicon
+        reyurns true if redirection was written or download can be skipped """
```
Comment (rgaudin) [inline sotoki/sotoki.py:994]:
```py
def handle_user_identicon(url, fullpath, convert_png, resize):
    """ Whether identicon is duplicate of existing (and process accordingly)
         
         Download identicon file if not (yet) a duplicate
         Write a redirection entry if a duplicate """
```

File: sotoki/sotoki.py [hunk 5]
```diff
@@ -890,13 +882,153 @@ def get_response_headers(url):
     return "default", "default"
 
 
-def download_image(url, fullpath, convert_png=False, resize=False):
+def get_redirection_to_duplicate_or_skip(url, convert_png, resize, ext):
```
Comment (rgaudin) [inline sotoki/sotoki.py:885]: I think we should invert the return code of this function. Would make better sense
```py
def get_identicon_shortcuts(url, convert_png, resize, ext):
    """" (skip, redirect_to) shortcuts for an identicon if possible
           
            skip (bool): whether this identicon is meaningless and should not be downloaded.
            redirect_to: URL of a copy of this identicon to redirect to """

File: sotoki/sotoki.py [hunk 6]
```diff
@@ -890,13 +882,153 @@ def get_response_headers(url):
...
+        # a specific image is duplicated, whatever the query part
+        if parsed_url.path == "/embedImage.png":
+            return (
+                download_and_return_name(
+                    url, "blend_exchange", convert_png, resize, ext
+                ),
+                False,
+            )
+
+    return None, None
```
Comment (rgaudin) [inline sotoki/sotoki.py:989]: Should thus be `return False, None`

**Review by rgaudin — CHANGES_REQUESTED** [review]
Couple of minor stuff

File: sotoki/sotoki.py (hunk 2 above)
Comment (satyamtg) [reply sotoki/sotoki.py:950]: These content-length values are the content-length value for the two types of default images, the default1 being the blue 'G' logo and default2 one being the green autogenerated identicon if hash is invalid these need to be manually changed if the images change but they are here for a pretty long time and shall stay to be so. I've put this comment there.

File: sotoki/sotoki.py (hunk 5 above)
Comment (satyamtg) [reply sotoki/sotoki.py:885]: Done

Comment (satyamtg) [reply sotoki/sotoki.py:885]: It's better if we call it get_image_shortcuts as this doesn't do it only for identicons

File: sotoki/sotoki.py (hunk 4 above)
Comment (satyamtg) [reply sotoki/sotoki.py:994]: Named it handle_duplicate_images() as it's not just for identicons, though most of the effect is on identicons

File: sotoki/sotoki.py (hunk 5 above)
Comment (rgaudin) [reply sotoki/sotoki.py:885]: OK then, right!

File: sotoki/sotoki.py (hunk 4 above)
Comment (rgaudin) [reply sotoki/sotoki.py:994]: Got it

File: sotoki/sotoki.py (hunk 2 above)
Comment (rgaudin) [reply sotoki/sotoki.py:957]: Can you add the URL of such default icon?

**Review by rgaudin — APPROVED** [review]

File: sotoki/sotoki.py (hunk 2 above)
Comment (satyamtg) [reply sotoki/sotoki.py:957]: Okay, I didn't read the comment, sorry. Here's a link to both of them -
https://www.gravatar.com/avatar/wrong_hash?s=128
https://www.gravatar.com/avatar/XXXXXXXXXXXXXXXXXXXXXXXXX?s=128&d=identicon

Will add them in the code whenever I push a new commmit.

Comment (rgaudin) [reply sotoki/sotoki.py:957]: pushed it to master
